### PR TITLE
[FEA][WIP] allow `DataFrame.join()` to take a list of DataFrames to merge with

### PR DESCRIPTION
**Pull request by KrishanBhasin**
I have a workflow that involves merging many (>100) dataframes with each other. I originally wrote a naive approach below, as Dask Dataframes don't fully match the Pandas API in allowing a list of dataframes being passed in.

```python
master_ddf = dd.read_parquet()  # this is the "driver" index
for ddf in lots_of_ddfs:
  master_ddf = master_ddf.join(ddf, how="left")
```
which was very slow and painful; the task graph looked like the below (created to avoid showing the generation of `lots_of_ddfs`):
image
with a task stream something like the below (for the avoidance of doubt, the left hand side the compute-heavy work of generating `lots_of_ddfs`, and the right hand side is all the join operations):
image

I implemented a pairwise approach to merging these hundreds of dataframes together which allowed it to be hugely parallelised, which I have started to generalise in this PR. The new task graph looks like the below:
image
and the new task stream no longer has the long tail to the right:
image

The major advantage for my specific use case was that once any given pair of dataframes finished being prepared/computing, their merges could start ahead of time; any dataframes that were slow to materialise don't have to hold up the entire merging process.

If there is positive feedback to this draft PR, I'd be happy to finish it off and add proper tests.

- [ ] Closes #xxxx
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask` / `isort dask`

cc @mmccarty

### CODE REVIEW

**Comment by jsignell** [comment]
Sorry this has sat for such a long time, but I just wanted to say that I like the idea of this a lot and would be interested in helping the work along. Do you have any questions about how to proceed?

**Comment by KrishanBhasin** [comment]
Thanks @jsignell! It look's like I've failed to push the tests I started setting up, so I'll do that as soon as I'm logged on to the other machine.

I was leaning to tests that simply prove that this function produces identical output to chained `.join()` calls, as I don't have much of an idea how I could write tests to verify the new structure of the task graph (or even if that is desired).

Something like:

```python
driver_df_loop = create_simple_driver_df()
for df in many_dfs:
  driver_df_loop = driver_df.merge(df)

driver_df_pairwise = create_simple_driver_df()
driver_df_pairwise = driver_df_pairwise.merge(many_dfs)

assert_frame_eq(driver_df_pairwise_, driver_df_loop)

```

Where I'd make the driver a few different lengths to prove the various edge cases

**Comment by KrishanBhasin** [comment]
Hey @jsignell, I think I'd actually appreciate a review with comments/direction.

The implementation is structurally complete but I'd appreciate direction on how to improve my defensive checks (I left `#TODO`s inline).

I'd also appreciate guidance wrt what kind of tests you'd like to see covering this, over the one I've got right now (I acknowledge the test itself is a bit scruffy)

File: dask/dataframe/tests/test_dataframe.py [hunk 1]
```diff
@@ -4668,3 +4668,62 @@ def test_dot_nan():
 
     assert_eq(s1.dot(s2), dask_s1.dot(dask_s2))
     assert_eq(s2.dot(df), dask_s2.dot(dask_df))
+
+
+# TODO: Can the setup here go into a fixture? or just a private function?
```
Comment (jsignell) [inline dask/dataframe/tests/test_dataframe.py:4673]: You should use `pytest.mark.parametrize` to check that various invalid `how` kwargs raise errors.

File: dask/dataframe/tests/test_dataframe.py [hunk 2]
```diff
@@ -4668,3 +4668,62 @@ def test_dot_nan():
...
+    dfs = [
+        dd.from_pandas(
+            pd.DataFrame({"a": [4, 5, 6], "b": [3, 2, 1]}, index=[5, 6, 8]), 3
+        ),
+        dd.from_pandas(
+            pd.DataFrame({"a": [7, 8, 9], "b": [0, 0, 0]}, index=[9, 9, 9]), 3
+        ),
+    ]
+
+    with pytest.raises(ValueError):
```
Comment (jsignell) [inline dask/dataframe/tests/test_dataframe.py:4687]: Use `match` to check the message as well.

File: dask/dataframe/tests/test_dataframe.py [hunk 3]
```diff
@@ -4668,3 +4668,62 @@ def test_dot_nan():
...
+        dfs_to_merge.append(ddf)
+
+    for ddf in dfs_to_merge:
+        ddf_loop = ddf_loop.join(ddf, how="left")
+
+    ddf_pairwise = dd.from_pandas(pd.DataFrame(index=[0, 1, 2, 3]), 3)
+
+    ddf_pairwise = ddf_pairwise.join(dfs_to_merge)
+
+    tm.assert_frame_equal(ddf_pairwise.compute(), ddf_loop.compute())
```
Comment (jsignell) [inline dask/dataframe/tests/test_dataframe.py:4729]: Use the dask.dataframe `assert_eq` function for a full check of the dask metdata as well as the computed values.

```suggestion
    assert_eq(ddf_pairwise, ddf_loop)
```

**Review by jsignell — COMMENTED** [review]
This is looking pretty good! It would be good to add some tests with dataframes that have different npartitions and ones with `lsuffix` and `rsuffix` - I'm not sure how those are supposed to work.

**Comment by KrishanBhasin** [comment]
I had not considered `lsuffix/`rsuffix` at all - when I used this it wasn't a concern I had to worry about. Let me have a think how to handle it (I'm also 100% open to suggestions!)

**Comment by jsignell** [comment]
Nice! It'd probably be worth checking what pandas does with suffixes in these cases. We can just follow their pattern probably.

**Comment by KrishanBhasin** [comment]
It looks like Pandas doesn't like it when the dataframes in your list have overlapping columns:

```python
# Create 2 identical dfs
dfs_to_merge = []
for i in range(2):
    df = pd.DataFrame(
        {
            f"A": [5, 6, 7, 8],
            f"B": [4, 3, 2, 1],
        },
        index=[0, 1, 2, 3],
    )
    dfs_to_merge.append(df)

# Create a base to join to
df_base = pd.DataFrame(index=[0, 1, 3])
    
df_base.join(dfs_to_merge, lsuffix="L", rsuffix="R")

>>> ValueError: Indexes have overlapping values: Index(['A', 'B'], dtype='object')
```

This is the same error currently thrown by the merge code in this PR when the columns overlap.

**Comment by jsignell** [comment]
> This is the same error currently thrown by the merge code in this PR when the columns overlap.

Nice! Ok if we are following pandas, then that's perfect.

**Review by jsignell — APPROVED** [review]
This looks great to me! Thanks for working on it @KrishanBhasin!

**Pull request merged**